client/src/components/Breadcrumb/Breadcrumb.js

```diff
@@ -25,34 +32,58 @@ class Breadcrumb extends Component {
     ));
   }
 
+  /**
+   * @returns {*}
+   */
   renderLastCrumb() {
     const crumb = this.getLastCrumb();
     if (!crumb) {
       return null;
     }
-
-    const iconClassNames = ['breadcrumb__icon'];
-    if (crumb.icon) {
-      iconClassNames.push(crumb.icon.className);
-    }
-
     return (
       <div className="breadcrumb__item breadcrumb__item--last">
         <h2 className="breadcrumb__item-title">
           {crumb.text}
-          {crumb.icon && (
-            <span
-              className={iconClassNames.join(' ')}
-              role="button"
-              tabIndex={0}
-              onClick={crumb.icon.onClick}
-            />
-          )}
+          {crumb.icon && this.renderIcons([crumb.icon])}
+          {crumb.icons && this.renderIcons(crumb.icons)}
         </h2>
       </div>
     );
   }
 
+  /**
+   * @param {Array} icon
+   * @returns {*}
+   */
+  renderIcons(icons) {
+    let k = 0;
+    return icons.map(icon => {
+      const { nodeName, className, hasRestrictedAccess, ...other } = icon;
+      // reassign with let so linter won't suggest 'const' above for unmodified nodeName/className
+      let attrs = { ...other };
+      const classNames = ['breadcrumb__icon'];
+      if (className) {
+        classNames.push(className);
+      }
+      attrs.className = classNames.join(' ');
+      attrs = { tabIndex: '0', ...attrs };
+      if (attrs.hasOwnProperty('onClick')) {
+        attrs = { role: 'button', ...attrs };
+      }
+      attrs.key = `breadcrumb-icon-${k}`;
```

Comment:
Array index (which you can get as the second argument on your map function) [isn't suitable as a key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318), even if you prefix it with a string. You should use something else that is a unique property on the icon if possible. Otherwise, you can avoid the `let k` by using the second arg of the mapping function.

Reply:
Updated to use lodash.uniqueid

Reply:
Is there no property in the `icon` object that's sufficiently unique? It doesn't have to be unique in the DOM, just in that array. It'd be nice to avoid adding the dependency if possible. Also, I'm not completely sure on the theory here, but I assume using unique ID is no better, as the key is changing on each re-render, which sort of ruins the point of it...

Reply:
I didn't find anything too overly unique on the icon object.  

>> It'd be nice to avoid adding the dependency if possible

Agree, though at the same time it's a tiny size import

>> as the key is changing on each re-render,

Yes that's what happens.  uniqueId() is just basically just i++.  

>> which sort of ruins the point of it

It does? The article you linked to implied the point of unique keys was to avoid collisions, which uniqueId() will avoid

Reply:
What's the likelihood of rendering the same icon twice in a single breadcrumb? It feels like that'd be a much better piece of information to use for an ID than a generated number, and I  agree with @ScopeyNZ that pulling in a dependency to achieve effectively the same approach as you were taking before isn't a good outcome.

Reply:
The article I linked has a key point:

> a key is the only thing React uses to identify DOM elements

The problem with an array index is that re-ordering items (for example) will change which item exists at each key without React actually seeing the change. The problem will manifest as React choosing not to re-render the re-order. By using `uniqueid` you're doing the opposite, the key still doesn't relate to the DOM element, but this time it's regenerated every time, meaning React can't optimise the render as it's going to re-render each key every time. Marginally better, but removing the point of `key`.

Reply:
It's a slightly different outcome than before.  For a breadcrumb with a single icon, previously the key was `breadcrumb-icon-1`, then if you navigated away from the folder, to another folder with a single icon, then back to the original folder, then then the key is still `breadcrumb-icon-1`.  Now when you do the above it'll be `breadcrumb-icon-3`

If there's a quick way to convert an object to something that can be used as key (e.g. md5(icon) then cool, though I suspect that would also require a third party library

Feels like we're talking about something that doesn't really matter a whole lot though, cos it seemed like it worked fine before.  And if it's inefficient now with uniqueid, it seems like trivial amount of inefficiency

Reply:
Ok, then I'd prefer we go back to my original suggestion, if there's nothing sufficiently unique, use the index, but extract it from the mapping function args rather than maintaining your own increment. It seems unlikely the icons are going to be re-ordered by the UI anyway right?

We want to always do things correctly in our repos btw, pointing out inefficiencies might have little impact now, but could end up being important down the line. And also, contributors are likely to use existing code as a reference, so we should be setting good examples.

Reply:
Has been changed to use the index from map()